spec/index.bs

```diff
@@ -7625,11 +7642,135 @@ dictionary GPURenderPassLayout: GPUObjectDescriptorBase {
+        tightly packed blocks of **four 32-bit unsigned integer values (16 bytes total)**, given in the same
+        order as the arguments for {{GPURenderEncoderBase/draw()}}. For example:
+
+        <pre highlight="js">
+            let drawIndirectParameters = new Uint32Array(8);
+            // first draw
+            drawIndirectParameters[0] = vertexCount0;
+            drawIndirectParameters[1] = instanceCount0;
+            drawIndirectParameters[2] = firstVertex0;
+            drawIndirectParameters[3] = firstInstance0;
+            // second draw
+            drawIndirectParameters[4] = vertexCount1;
+            drawIndirectParameters[5] = instanceCount1;
+            drawIndirectParameters[6] = firstVertex1;
+            drawIndirectParameters[7] = firstInstance1;
+        </pre>
+
+        The <dfn dfn for=>indirect multiDraw count</dfn> encoded in the |drawCountBuffer| must be a **32-bit unsigned
+        integer** giving the number of draws to issue from the |indirectBuffer|.
+
+        <div algorithm="GPURenderEncoderBase.multiDrawIndirect">
+            **Called on:** {{GPURenderEncoderBase}} this.
+
+            **Arguments:**
+            <pre class=argumentdef for="GPURenderEncoderBase/multiDrawIndirect(indirectBuffer, indirectOffset, maxDrawCount, drawCountBuffer, drawCountOffset)">
+                |indirectBuffer|: Buffer containing the [=indirect multiDraw parameters=].
+                |indirectOffset|: Offset in bytes into |indirectBuffer| where the drawing data begins.
+                |maxDrawCount|: Maxumum number of draws or exact number of draws if |drawCountBuffer| is `null`.
+                |drawCountBuffer|: Buffer containing the [=indirect multiDraw count=].
+                |drawCountOffset|: Offset in bytes into |drawCountBuffer| where the [=indirect multiDraw count=] is located.
+            </pre>
+
+            **Returns:** {{undefined}}
+
+            Issue the following steps on the [=Device timeline=] of |this|.{{GPUObjectBase/[[device]]}}:
+            <div class=device-timeline>
+                1. If any of the following conditions are unsatisfied, make |this| [=invalid=] and stop.
+                    <div class=validusage>
+                        - It is [$valid to draw$] with |this|.
+                        - |indirectBuffer| is [$valid to use with$] |this|.
+                        - |indirectBuffer|.{{GPUBuffer/[[usage]]}} contains {{GPUBufferUsage/INDIRECT}}.
+                        - |indirectOffset| + sizeof([=indirect multiDraw parameters=]) &le;
+                            |indirectBuffer|.{{GPUBuffer/[[size]]}}.
+                        - |indirectOffset| is a multiple of 4.
+                        - |maxDrawCount| &le; limits.{{supported limits/maxMultiDrawIndirectCount}}.
+                        - |drawCountBuffer| is `null`, unless the {{GPUFeatureName/"multi-draw-indirect"}} [=feature=] is enabled.
```

Comment:
the place above this says:
> This method is defined if and only if the {{GPUFeatureName/"multi-draw-indirect"}} [=feature=] is enabled.

if the method isn't defined without a feature, why does it support nullable draw count buffer?

Reply:
This is replicating how both Vulkan and DirectX behave, if the `drawCountBuffer` is null then `maxDrawCount` becomes just the draw count. In some cases the number of draws will be known CPU side.

Reply:
If we know the draw count on CPU, we can issue the individual draw calls ourselves, too. And the user can do it.
A few questions:
  - does this work consistently on all APIs, or would we need a polyfill somewhere?
  - should the draw call itself be available in core, and only the `drawCountBuffer` requiring the feature?
  - is this generally useful? (I assume yes, just bringing it for completeness)

Reply:
Thank you for elaboration! I wasn't clear enough in my questions, and I was missing a bit of context here. I understand better now :)
How much is the whole thing useful *considering* we don't expose `gl_DrawID` kind of thing? I imagine the numbers you are providing are for the case which does know the draw ID. Also, could you point me to the piece of investigation showing why we can't have that `draw_index` builtin?

Reply:
I did not consider a `draw_index` in the initial investigation (as I was avoiding a WGSL component) and I don't know enough about HLSL to know for D3D12 but on Vulkan it's the `VK_KHR_shader_draw_parameters` extension on v1.0 (with 86% support) and a core feature on v1.1 (with 99.5% support).

While it would definitely be nice, and could easily be folded into this feature (assuming there is a D3D12 equivalent), it's not required as long as `first-instance-indirect` is supported. This is because when you can set `firstInstance` to be non-zero you can use `instance_index` (when not using instance stride vertex data) to encode both the instance and draw ID, which was the workaround on Vulkan before the extension.